## Scipy and Optimization

In the last notebook we took a tour of the `numpy` and `scipy` stack as it relates to math and statistics. This notebook is going to focus on another subset of `scipy` functionality that is absolutely critical to data analytics. Nearly every statistical algorithm aside from Ordinary Least Squares (OLS) regression cannot be solved algebraically. Because they cannot be solved algebraically, they must be solved **numerically**.

### Numeric Optimization

Let's start with an example of optimization. Let's say that you know the demand function for tickets to watch a local sports franchise. You can write the inverse demand function as

$$ P = 300 - \frac{1}{2}Q $$

and the total cost function as

$$ TC = 4000 + 45Q $$

In order to choose the right number of tickets to sell, you need to calculate the quantity of tickets that will maximize profits. We can calculate total revenue as $ TR = P \times Q $, and we can calculate profit as $ \Pi = TR - TC $. This means that our profit function is

$$ \Pi = 300Q - \frac{1}{2}Q^2 - 4000 - 45Q $$

In order to find the $Q$ associated with the highest achievable level of profit, we can use calculus to find the point at which the rate of change in the profit function is zero ($\frac{\partial\Pi}{\partial Q}=0$).

$$ \frac{\partial\Pi}{\partial Q} = 300 - Q - 45 = 0 \implies Q = 255$$

So we can **algebraically** solve this particular problem. This isn't always the case. Using `scipy`, we can solve this same problem, as well as many algebraically intractable problems that might be more interesting to us.


<img src="https://github.com/dustywhite7/Econ8320/raw/master/SlidesCode/paraboloid.png" width="200" height="200" />


In any optimization problem, we need to find a way to get ourselves to the minimum, and to know when we get there. When we look at the above image, we are able to visually trace the functional shape (looks like a rainbow ice cream cone to me...) and locate the bottom of the function. What we want to do is utilize an algorithm to "trace" our way from an arbitrary starting point within a function to the optimal point in that function.

In three or fewer dimensions, this is easy. Regressions and statistical models often live in worlds with 100's or 1000's (even millions sometimes) of dimensions. We can't visualize our way to the bottom of those functions!

The class of algorithm that is used to solve these problems is called **gradient descent**.

<img src="https://github.com/dustywhite7/Econ8320/raw/master/SlidesCode/gradDesc.png" width="400" />

**Gradient descent** is an algorithm that explores the shape of the function, and determines which direction is most likely to lead to the optimal point. Let's focus on minimization. We want to find our way to the *bottom* of a function, and we can use gradient descent to try to get there. Given any starting point, our goal is to check the direction in which we can move downward most quickly, and start moving in that direction. At some distance from our starting point, we will stop and re-evaluate the direction in which we would like to travel. Are we still heading downhill in the steepest direction? If we aren't, then we need to update our behavior.

Our gradient descent algorithm will keep "looking around" and moving down until it reaches a point at which it can no longer move "down" in any meaningful way. That is the stopping point, and is treated as the optimum.


With an intuitive understanding of how optimization will happen computationally, it's time to learn a bit more about the math and the code that will help us to achieve computational optimization.

Consider a function, $f$, with two variables $x$ and $y$. Because there are two input variables in the function, it has two partial derivatives:

$$ \frac{\partial f}{\partial x} \text{ and } \frac{\partial f}{\partial y} $$

Each partial derivative tells us how $f$ changes as we move in a particular dimension **all else equal**. The **gradient**, then, is the vector of all partial derivatives of a given function at any point along the function:

$$ \nabla f = \left[ \begin{matrix} \frac{\partial f}{\partial x} \\ \\ \frac{\partial f}{\partial y} \end{matrix} \right]  $$

We can use the gradient to determine the linear approximation of a function at any given point. Think about the gradient as the mathematical representation of the slope and direction of a hill you are hiking on. If we know the gradient, we know which way is down. As we continue to calculate gradients while walking, we can continue to ensure that we will walk downhill until we reach the bottom of the hill.




The steps of our gradient descent function will be the following:

- Evaluate the gradient of the function
- Find the direction of steepest descent
- Determine how far to move in that direction
- Move to new point
- Reevaluate the gradient
- Stop moving when gradient is within a margin of error from 0

Let's try to implement gradient descent by solving our old profit maximization problem computationally. The very first thing that we need to do is write a Python function that will represent our mathematical function.

In [28]:
def profit(q):
    p = 300-0.5*q
    tr = p*q
    tc = 4000 + 45*q
    return tr - tc

This function will allow us to calculate profit at any output level based on our assumed total costs and demand curve. With this function, we can quickly calculate the gradient (in this case, just a simple derivative because our function is univariate) by calculating profit at two nearby points, and dividing by the horizontal distance between those points:

In [29]:
# Gradient at q=200

(profit(201) - profit(199))/2

55.0

    55.0



Thus, a one unit increase in output at $Q=200$ results in a $55 increase in profits. This is cool, but it isn't enough for us to find the point of maximum profit (the optimal point). For that, we will need to calculate LOTS of gradients in order to move along the function until we cannot increase profits any further.

Fortunately for us, `scipy` comes with optimization tools that will do all of the heavy lifting of the "search" for the optimal point. All that we need to do is frame our question algorithmically, and let `scipy` do the rest:

In [30]:
import numpy as np
from scipy.optimize import minimize

We start by importing the `minimize` function from `scipy.optimize`. Hang on! Weren't we working on a MAXIMIZATION problem?? What are we doing here?

Maximization and minimization are the **same thing**. To maximize a function, you can multiply that function by `-1` and then calculate the minimum of the new "upside-down" function. It is functionally equivalent. So, in computational optimization, we always minimize.

### Prepping for optimization

As we prepare to optimize, there are two common problems with our function that we may need to resolve:

1) When using `minimize` we can only pass an array of inputs, so we have to be careful to write our function accordingly
2) Our problem is concave, and so has a maximum
	- We need to restate it as a minimization problem

Problem 1 does not apply, since our function in univariate. In order to make our problem a minimization problem, we can flip our profit maximization function. We will simply return -1 * profit:

In [31]:
def profit(q):
    p = 300-0.5*q
    tr = p*q
    tc = 4000 + 45*q
    pi =  tr - tc # didn't name it profit because that is our function's name! Don't want to clutter our name space!
    return -1*pi

### Making the call to `minimize`

Now that our function is ready, it is time to minimize! The `minimize` function takes two arguments:
1. Our function that we want to optimize
2. A starting guess (as a vector)

Let's give it a try.

In [32]:
res = minimize(profit, [0]) # provide function and starting inputs
res

  message: Optimization terminated successfully.
  success: True
   status: 0
      fun: -28512.499999980355
        x: [ 2.550e+02]
      nit: 3
      jac: [ 0.000e+00]
 hess_inv: [[ 1.000e+00]]
     nfev: 14
     njev: 7

          fun: -28512.499999980355
     hess_inv: array([[1.00000175]])
          jac: array([0.])
      message: 'Optimization terminated successfully.'
         nfev: 21
          nit: 3
         njev: 7
       status: 0
      success: True
            x: array([255.00019821])



That's it! No calculus, no searching, no checking gradients manually. `minimize` simply takes our function and our starting guess and brings us back the optimal choice. We get lots of information stored in the attributes of the `res` object:

- `fun` provides the value of the function (this is -1 times the profit level at the optimal output in our example)
- `hess_inv` and `jac` are measures of gradient and are used to determine how far to go and in which direction
- `message` should be self-explanatory
- `nfev` is the number of times the function (in this case `profit`) was evaluated during the search
- `nit` is the number of iterations it took to find the optimum
- `njev` is the number of times the Jacobian was estimated
- `status` is a code associated with the `message` and `success` atrributes
- `success` tells you whether or not an optimum was found (sometimes it cannot be easily found!)
- `x` probably the most important attribute. This tells us the optimal input value (in our case $Q$), or set of values depending on our function. In a regression context, this could represent the fitted coefficients!

Going forward, you will realize (especially because so many of them print the output as they optimize) just how many libraries in Python use this minimize function behind the scenes. It is used in `statsmodels`, `sklearn`, and many other high-profile libraries! Now that you know where it really happens (in `scipy`!), you'll be better able to troubleshoot the problems that will inevitably arise as you use statistical models.

**Solve-it!**

In this lesson we learned about optimization using SciPy. For the assignment this week, I would like you to build off of your `RegressionModel` class. You will add a Logistic Regression (Logit) method to your class, so that when the `regression_type` parameter is `logit`, Logistic Regression Results are returned.

Your job is to create the following functionality within your class object:
- a method (call it `logistic_regression`) that estimates the results of logistic regression using your `x` and `y` data frames, and using a likelihood function and gradient descent (DO NOT USE PREBUILT REGRESSION FUNCTIONS).
    - You need to write a function to calculate the Log-likelihood of your model
    - You need to implement gradient descent to find the optimal values of beta
    - You need to use your beta estimates, the model variance, and calculate the standard errors of the coefficients, as well as Z statistics and p-values
    - the results should be stored in a dictionary named `results`, where each variable name (including the intercept if `create_intercept` is `True`) is the key, and the value is another dictionary, with keys for `coefficient`, `standard_error`, `z_stat`, and `p_value`. The coefficient should be the log odds-ratio (which takes the place of the coefficients in OLS)
- a method called `fit_model` that uses the `self.regression_type` attribute to determine whether or not to run an OLS or Logistic Regression using the data provided. This method should call the correct regression method.
- an updated method (call it `summary`) that presents your regression results in a table
    - Columns should be: Variable name, Log odds-ratio value, standard error, z-statistic, and p-value, in that order.
    - Your summary table should have different column titles for OLS and Logistic Regressions! (think if statement...)

You only need to define the class. My code will create an instance of your class (be sure all the names match these instructions, and those from last week!), and provide data to run a regression. I will provide the same data to you, so that you can experiment and make sure that your code is functioning properly.

NOTE: I have created a [primer on Logistic regression](https://github.com/dustywhite7/Econ8320/blob/master/SlidesPDF/9-2%20-%20Logit%20Primer.pdf) to go with this assignment. See the Github slidesPDF folder

Put the code that you would like graded in the cell labeled `#si-exercise`. I recommend copying your code from the last assignment (in chapter 9 about linear regression and `numpy`), and continuing from there.



In [33]:
#si-exercise
import numpy as np
import pandas as pd
import scipy.stats as stat
from scipy.optimize import minimize

class RegressionModel:
    def __init__(self, x, y, create_intercept = True, regression_type = 'ols'):
        self.x = x
        self.y = y
        self.create_intercept = create_intercept
        self.regression_type = regression_type
        self.results = {}
        self.add_intercept()
        self.fit_model()
        print(self.summary())

    def add_intercept(self):
        if self.create_intercept and 'intercept' not in self.x.columns:
            self.x['intercept'] = 1

    def ols_regression(self):
        # Convert dataframes to numpy arrays
        X = self.x.values
        y = self.y.values

        # Calculate OLS coefficients
        # (X^T * X)^(-1) * X_T * y
        X_T = X.T
        XTX = X_T @ X
        XTX_inv = np.linalg.inv(XTX)
        XTy = X_T @ y
        coefficients = XTX_inv @ XTy

        # Calculate predicted values and residuals
        y_pred = X @ coefficients
        residuals = y - y_pred

        # Calculate Residual Sum of Squares (RSS)
        rss = np.sum(residuals**2)

        # Calculate degrees of freedom
        n = X.shape[0] # number of observations
        k = X.shape[1] # number of predictors (including intercept)
        df = n - k

        # Calculate Mean Squared Error (MSE)
        mse = rss / df

        # Calculate covariance matrix of coefficients
        covariance_matrix = mse * XTX_inv

        # Extract standard errors from the diagonal of the covariance matrix
        standard_errors = np.sqrt(np.diag(covariance_matrix))

        # Calculate t-statistic
        t_stats = coefficients / standard_errors

        # Calculate p-values (one-tailed)
        p_values = 2 * (1 - stat.t.cdf(np.abs(t_stats), df)) # Use absolute value for two-tailed p-value

        # Store results in the self.results dictionary
        variable_names = self.x.columns
        for i, var_name in enumerate(variable_names):
            self.results[var_name] = {
                'coefficient': coefficients[i],
                'standard_error': standard_errors[i],
                't_stat': t_stats[i],
                'p_value': p_values[i]
            }

    def log_transform(self, x):
        # Logistic transformation (sigmoid function)
        return 1 / (1 + np.exp(-x))


    def logistic_regression(self):
        X = self.x.values
        y = self.y.values.flatten() # Flatten y for logistic regression

        # Define the negative log-likelihood function
        def neg_log_likelihood(beta):
            linear_model = X @ beta
            # Use the log_transform method for the sigmoid
            sigmoid = self.log_transform(linear_model)
            # Ensure numerical stability for the log
            sigmoid = np.clip(sigmoid, 1e-15, 1 - 1e-15)
            # The negative log-likelihood formula for logistic regression
            return -np.sum(y * np.log(sigmoid) + (1 - y) * np.log(1 - sigmoid))


        # Set initial guess for beta (zeros) - this is generally a good starting point
        initial_beta = np.zeros(X.shape[1])

        # Minimize the negative log-likelihood to get the coefficients
        try:
            res = minimize(neg_log_likelihood, initial_beta, method='BFGS')

            # Check for successful optimization
            if not res.success:
                print(f"Warning: Optimization did not converge successfully. Message: {res.message}")


            # Extract coefficients obtained from minimization
            coefficients = res.x

            # --- Calculate variance/covariance matrix using the modified OLS method as per hint ---
            # Calculate s-hat squared as n * y_bar * (1 - y_bar)
            n = X.shape[0] # number of observations
            y_bar = np.mean(y) # average value of y
            s_hat_squared = n * y_bar * (1 - y_bar)

            # Ensure s_hat_squared is non-negative for numerical stability, although for binary y, y_bar is between 0 and 1, so this should be fine.
            s_hat_squared = max(s_hat_squared, 1e-15)

            # Calculate (X^T * X)^(-1)
            X_T = X.T
            XTX = X_T @ X
            try:
                XTX_inv = np.linalg.inv(XTX)
            except np.linalg.LinAlgError:
                print("Warning: Could not compute inverse of XTX for standard error calculation.")
                XTX_inv = np.full((X.shape[1], X.shape[1]), np.nan) # Fill with NaN if inverse fails

            # Calculate the covariance matrix using the modified OLS formula: s_hat_squared * (X^T * X)^(-1)
            covariance_matrix = s_hat_squared * XTX_inv

            # --- End of modified OLS method for variance/covariance matrix ---


            # Extract standard errors from the diagonal of the covariance matrix
            if covariance_matrix is not None:
                 # Ensure diagonal elements are non-negative before taking sqrt
                 diag_covariance = np.diag(covariance_matrix)
                 if np.any(diag_covariance < 0):
                      print("Warning: Negative values in covariance matrix diagonal. Standard errors may be inaccurate.")
                      standard_errors = np.sqrt(np.maximum(diag_covariance, 0)) # Take sqrt of non-negative part
                 else:
                      standard_errors = np.sqrt(diag_covariance)
            else:
                 standard_errors = np.full(len(coefficients), np.nan) # Fill with NaN if covariance matrix is not available


            # Calculate z-statistics
            # Avoid division by zero if standard_errors contains zero or near-zero values
            z_stats = np.divide(coefficients, standard_errors, out=np.full_like(coefficients, np.nan), where=standard_errors!=0)


            # Calculate p-values (two-tailed)
            p_values = 2 * (1 - stat.norm.cdf(np.abs(z_stats)))


            # Store results
            variable_names = self.x.columns
            for i, var_name in enumerate(variable_names):
                self.results[var_name] = {
                    'coefficient': coefficients[i],
                    'standard_error': standard_errors[i],
                    'z_stat': z_stats[i],
                    'p_value': p_values[i]
                }

        except Exception as e:
            print(f"An error occurred during logistic regression: {e}")
            # Optionally, store an error message in results or raise the exception
            self.results = {"Error": f"Could not compute logistic regression: {e}"}



    def fit_model(self):
        if self.regression_type == 'ols':
            self.ols_regression()
        elif self.regression_type == 'logit':
            self.logistic_regression()
        else:
            raise ValueError("Invalid regression_type. Choose 'ols' or 'logit'.")


    def summary(self):
        summary_data = []
        if self.regression_type == 'ols':
            # Add intercept row first if it exists
            if 'intercept' in self.results:
                 stats = self.results['intercept']
                 summary_data.append({
                    "Variable Name": 'intercept',
                    "Coefficient": stats['coefficient'][0] if isinstance(stats['coefficient'], np.ndarray) else stats['coefficient'],
                    "Std. Error": stats['standard_error'][0] if isinstance(stats['standard_error'], np.ndarray) else stats['standard_error'],
                    "t-statistic": stats['t_stat'][0] if isinstance(stats['t_stat'], np.ndarray) else stats['t_statistic'],
                    "P-value": stats['p_value'][0] if isinstance(stats['p_value'], np.ndarray) else stats['p_value']
                })
            for var_name, stats in self.results.items():
                if var_name != 'intercept': # Add other variables
                    summary_data.append({
                        "Variable Name": var_name,
                        "Coefficient": stats['coefficient'][0] if isinstance(stats['coefficient'], np.ndarray) else stats['coefficient'],
                        "Std. Error": stats['standard_error'][0] if isinstance(stats['standard_error'], np.ndarray) else stats['standard_error'],
                        "t-statistic": stats['t_stat'][0] if isinstance(stats['t_stat'], np.ndarray) else stats['t_statistic'],
                        "P-value": stats['p_value'][0] if isinstance(stats['p_value'], np.ndarray) else stats['p_value']
                    })
            summary_df = pd.DataFrame(summary_data, columns=["Variable Name", "Coefficient", "Std. Error", "t-statistic", "P-value"])
        elif self.regression_type == 'logit':
             # Check if results contain an error message
            if "Error" in self.results:
                 summary_df = pd.DataFrame([{"Error": self.results["Error"]}])
            else:
                # Add intercept row first if it exists
                if 'intercept' in self.results:
                    stats = self.results['intercept']
                    # Explicitly set p-value to 0 for intercept if very small
                    intercept_p_value = stats['p_value']
                    if intercept_p_value < 1e-9: # Using a small threshold to consider it effectively zero
                        intercept_p_value = 0.0
                    summary_data.append({
                        "Variable Name": 'intercept',
                        "Log Odds-Ratio": stats['coefficient'],
                        "Std. Error": stats['standard_error'],
                        "z-statistic": stats['z_stat'],
                        "P-value": intercept_p_value
                    })
                for var_name, stats in self.results.items():
                     if var_name != 'intercept': # Add other variables
                        summary_data.append({
                            "Variable Name": var_name,
                            "Log Odds-Ratio": stats['coefficient'],
                            "Std. Error": stats['standard_error'],
                            "z-statistic": stats['z_stat'],
                            "P-value": stats['p_value']
                        })
                summary_df = pd.DataFrame(summary_data, columns=["Variable Name", "Log Odds-Ratio", "Std. Error", "z-statistic", "P-value"])
        else:
            summary_df = pd.DataFrame(["No results to display. Fit a model first."])
        return summary_df